### PR TITLE
ENH add sample #2419

### LINKED ISSUE #2419: Series/DataFrame sample method with/without replacement
Should use a more intelligent algorithm than using `np.random.permutation`

### PR DESCRIPTION
**Pull request by hayd**
fixes #2419

Hmmm `np.random.choice` not available on numpy < 1.7.

### CODE REVIEW

**Comment by jreback** [comment]
closing for now. @hayd if you want to update at some point, ok,.

File: pandas/stats/misc.py [hunk 1]
```diff
@@ -297,3 +297,20 @@ def _bucket_labels(series, k):
         mat[v] = i
 
     return mat + 1
+
+
+def choice(arr, size, replace):
```
Comment (shoyer) [inline pandas/stats/misc.py]: we don't need this anymore since we require numpy 1.7 now

File: pandas/core/generic.py [hunk 2]
```diff
@@ -1241,6 +1241,45 @@ def take(self, indices, axis=0, convert=True, is_copy=True):
...
+        If you are sampling without replacement over a larger sample size than
+        the object you're sampling a ValueError will be raised.
+
+        """
+        try:
+            from numpy.random import choice
+        except ImportError:
+            from pandas.stats.misc import choice
+        msk = choice(len(self), size, replace=replace)
+        return self.iloc[msk]
```
Comment (shoyer) [inline pandas/core/generic.py]: This should use `take`, which is faster

**Comment by shoyer** [comment]
Just wrote a little utility function to do this, so I may pick this up.

```python
def sample_n(df, size, replace=False, weight=None, seed=None):
    rs = np.random.RandomState(seed)
    locs = rs.choice(df.shape[0], size=size, replace=replace, p=weight)
    return df.take(locs, axis=0)
```

**Pull request closed without merging**